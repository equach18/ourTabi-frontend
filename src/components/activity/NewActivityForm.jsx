import { useState } from "react";
import { useNavigate } from "react-router-dom";
import { useContext } from "react";
import TripContext from "../../context/TripContext";
import Alert from "../common/Alert";

const validCategories = [
  "food",
  "hiking",
  "tours",
  "shopping",
  "adventure",
  "outdoors",
  "other",
];

function NewActivityForm() {
  const { addActivity } = useContext(TripContext);
  const navigate = useNavigate();

  const [formData, setFormData] = useState({
    name: "",
    category: "other",
    description: "",
    location: "",
    scheduledTime: "",
  });

  const [formErrors, setFormErrors] = useState([]);
  const [isSubmitting, setIsSubmitting] = useState(false);

  async function handleSubmit(evt) {
    evt.preventDefault();
    setIsSubmitting(true);
    setFormErrors([]);

    const formattedScheduledTime = formData.scheduledTime
      ? new Date(formData.scheduledTime).toISOString()
      : undefined;

    const activityToSubmit = {
      ...formData,
      scheduledTime: formattedScheduledTime,
    };

    try {
      await addActivity(activityToSubmit);
      navigate(-1);
    } catch (err) {
      setFormErrors(err);
    } finally {
      setIsSubmitting(false);
    }
  }

  /** Update form data field */
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((prev) => ({ ...prev, [name]: value }));
  }

  return (
    <div className="h-[calc(100vh-4rem)] flex items-center justify-center px-4 py-10">
      <div className="w-full max-w-xl bg-white p-8 rounded-xl shadow-lg">
        <h2 className="text-3xl font-bold text-gray-600 mb-4">
          Add a New Activity
        </h2>

        <form onSubmit={handleSubmit} className="space-y-3">
          {/* error alerts */}
          <Alert type="error" messages={formErrors} />

          {/* name */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="name">
              Activity Name
            </label>
            <input
              type="text"
              name="name"
              id="name"
              value={formData.name}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              required
            />
          </div>

          {/* category */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="category">
              Category
            </label>
            <select
              name="category"
              id="category"
              value={formData.category}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            >
              {validCategories.map((cat) => (
                <option key={cat} value={cat}>
                  {cat.charAt(0).toUpperCase() + cat.slice(1)}
                </option>
              ))}
            </select>
          </div>

          {/* description */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="description">
              Description (optional)
            </label>
            <textarea
              name="description"
              id="description"
              value={formData.description}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              maxLength={500}
            />
          </div>

          {/* location */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="location">
              Location (optional)
            </label>
            <input
              type="text"
              name="location"
              id="location"
              value={formData.location}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              maxLength={255}
            />
          </div>

          {/* scheduled time */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="scheduledTime">
              Scheduled Time (optional)
            </label>
            <input
              type="datetime-local"
              name="scheduledTime"
              id="scheduledTime"
              value={formData.scheduledTime}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            />
          </div>

          {/* Submit Button */}
          <button
            type="submit"
            className={`w-full bg-blue-500 text-white py-2 rounded-lg hover:bg-blue-600 transition ${
              isSubmitting && "opacity-50 cursor-not-allowed"
            }`}
            disabled={isSubmitting}
          >
            {isSubmitting ? "Adding..." : "Add Activity"}
          </button>
        </form>
      </div>
    </div>
  );
}

export default NewActivityForm;
