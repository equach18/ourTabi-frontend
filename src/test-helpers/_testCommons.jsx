import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../context/UserContext";
import TripContext from "../context/TripContext";

export const testUser1 = {
  id: 1,
  firstName: "Test1",
  lastName: "User1",
  username: "testUser1",
  profilePic: "",
};
export const testUser2 = {
  id: 2,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser2",
  profilePic: "testUser2.png",
};
export const testUser3 = {
  id: 3,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser3",
  profilePic: "testUser3.png",
};

export const defaultMembers = [
  {
    userId: testUser1.id,
    username: testUser1.username,
    profilePic: testUser1.profilePic,
  },
  {
    userId: testUser2.id,
    username: testUser2.username,
    profilePic: testUser2.profilePic,
  },
];

export const testActivity = {
  id: 1,
  name: "Kayaking",
  tripId: 53,
  category: "Other",
  location: "Chicago",
  description: "So much fun",
  scheduledTime: "2025-07-10T14:30:00Z",
};

export const testComment = {
  id: 1,
  text: "Test comment",
  createdAt: "2025-07-11T14:30:00Z",
  userId: 1,
};

export const testTrip = {
  id: 42,
  title: "Test trip",
  destination: "Chicago, IL",
  startDate: "2025-07-10",
  endDate: "2025-07-12",
  isPrivate: true,
};

export const testFriendsData = {
  friends: [testUser1 ],
  incomingRequests: [testUser2],
  sentRequests: [testUser3],
};

// Default mock context values
const defaultUserContext = { currentUser: testUser1 };
const defaultTripContext = {
  voteOnActivity: vi.fn(),
  votes: {},
  deleteComment: vi.fn(),
  addMember: vi.fn(),
  removeMember: vi.fn(),
  members: defaultMembers,
};

// create mock trip context
export function createMockTripContext(overrides = {}) {
  return {
    voteOnActivity: vi.fn(),
    votes: {},
    deleteComment: vi.fn(),
    members: [],
    addMember: vi.fn(),
    removeMember: vi.fn(),
    ...overrides,
  };
}

/**
 * Renders a component inside MemoryRouter
 */
export function renderWithRouter(ui) {
  return render(<MemoryRouter>{ui}</MemoryRouter>);
}

/**
 * Renders a component inside MemoryRouter + UserContext + TripContext
 */
export function renderWithAllProviders(
  ui,
  {
    userContext = defaultUserContext,
    tripContext = defaultTripContext,
    route = "/",
  } = {}
) {
  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserContext.Provider value={userContext}>
        <TripContext.Provider value={tripContext}>{ui}</TripContext.Provider>
      </UserContext.Provider>
    </MemoryRouter>
  );
}

/**
 * Renders a component with just UserContext
 */
export function renderWithUserProvider(ui, { userContext } = {}) {
  return render(
    <UserContext.Provider value={userContext}>{ui}</UserContext.Provider>
  );
}
