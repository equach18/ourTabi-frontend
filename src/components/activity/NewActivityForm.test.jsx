import { fireEvent, screen, waitFor } from "@testing-library/react";
import { Routes, Route } from "react-router-dom";
import NewActivityForm from "./NewActivityForm";
import { renderWithAllProviders } from "../../test-helpers/_testCommons";
// mock useNavigate
const mockNavigate = vi.fn();
vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});
describe("NewActivityForm", () => {
  const mockAddActivity = vi.fn();

  const tripContextValue = {
    addActivity: mockAddActivity,
  };

  function renderForm() {
    return renderWithAllProviders(
      <Routes>
        <Route
          path="/trips/:tripId/activities/new"
          element={<NewActivityForm />}
        />
      </Routes>,
      {
        tripContext: tripContextValue,
        route: "/trips/1/activities/new",
      }
    );
  }

  beforeEach(() => {
    mockAddActivity.mockClear();
    mockNavigate.mockClear();
  });

  it("matches snapshot", () => {
    const { asFragment } = renderForm();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders form inputs", () => {
    renderForm();

    expect(screen.getByLabelText(/activity name/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/category/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/description/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/location/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/scheduled time/i)).toBeInTheDocument();
  });

  it("calls addActivity when form is submitted", async () => {
    renderForm();

    fireEvent.change(screen.getByLabelText(/activity name/i), {
      target: { value: "pickleball" },
    });

    fireEvent.change(screen.getByLabelText(/category/i), {
      target: { value: "outdoors" },
    });

    fireEvent.click(screen.getByRole("button", { name: /add activity/i }));

    await waitFor(() => {
      expect(mockAddActivity).toHaveBeenCalledWith(
        expect.objectContaining({
          name: "pickleball",
          category: "outdoors",
        })
      );
    });
  });

  it("displays error messages if addActivity fails", async () => {
    mockAddActivity.mockRejectedValue(["a mock error!!"]);
    renderForm();

    fireEvent.change(screen.getByLabelText(/activity name/i), {
      target: { value: "wrong activity" },
    });

    fireEvent.click(screen.getByRole("button", { name: /add activity/i }));

    await waitFor(() => {
      expect(screen.getByText(/a mock error/i)).toBeInTheDocument();
    });
  });

  it("navigate(-1) is called after successful submission", async () => {
    mockAddActivity.mockResolvedValue();
    renderForm();

    fireEvent.change(screen.getByLabelText(/activity name/i), {
      target: { value: "Swimming" },
    });

    fireEvent.click(screen.getByRole("button", { name: /add activity/i }));

    await waitFor(() => {
      expect(mockAddActivity).toHaveBeenCalled();
      expect(mockNavigate).toHaveBeenCalledWith(-1); // âœ… This tests navigate(-1)
    });
  });
});
