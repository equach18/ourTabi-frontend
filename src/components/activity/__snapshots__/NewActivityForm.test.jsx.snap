// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewActivityForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-[calc(100vh-4rem)] flex items-center justify-center px-4 py-10"
  >
    <div
      class="w-full max-w-xl bg-white p-8 rounded-xl shadow-lg"
    >
      <h2
        class="text-3xl font-bold text-gray-600 mb-4"
      >
        Add a New Activity
      </h2>
      <form
        class="space-y-3"
      >
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="name"
          >
            Activity Name
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="name"
            name="name"
            required=""
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="category"
          >
            Category
          </label>
          <select
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="category"
            name="category"
          >
            <option
              value="food"
            >
              Food
            </option>
            <option
              value="hiking"
            >
              Hiking
            </option>
            <option
              value="tours"
            >
              Tours
            </option>
            <option
              value="shopping"
            >
              Shopping
            </option>
            <option
              value="adventure"
            >
              Adventure
            </option>
            <option
              value="outdoors"
            >
              Outdoors
            </option>
            <option
              value="other"
            >
              Other
            </option>
          </select>
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="description"
          >
            Description (optional)
          </label>
          <textarea
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="description"
            maxlength="500"
            name="description"
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="location"
          >
            Location (optional)
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="location"
            maxlength="255"
            name="location"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="scheduledTime"
          >
            Scheduled Time (optional)
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="scheduledTime"
            name="scheduledTime"
            type="datetime-local"
            value=""
          />
        </div>
        <button
          class="w-full bg-blue-500 text-white py-2 rounded-lg hover:bg-blue-600 transition false"
          type="submit"
        >
          Add Activity
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
